pkg/terminal/proxy.go

```diff
@@ -0,0 +1,175 @@
+			}
+		}
+	}
+
+	ws, err := client.Resource(WorkspaceGroupVersionResource).Namespace(namespace).Get(workspaceName, metav1.GetOptions{})
+	if err != nil {
+		http.Error(w, "Failed to get the requested workspace. Cause: "+err.Error(), http.StatusForbidden)
+		return
+	}
+
+	creator := ws.GetAnnotations()["org.eclipse.che.workspace/creator"]
+	if creator != userId {
+		http.Error(w, "User is not a owner of the requested workspace", http.StatusForbidden)
+		return
+	}
+
+	terminalHost, err := p.getBaseTerminalHost(ws)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	p.proxyToWorkspace(terminalHost, path, user.Token, r, w)
+}
+
+//stripTerminalAPIPrefix strips path prefix that is expected for Terminal API request
+func stripTerminalAPIPrefix(requestPath string) (ok bool, namespace string, workspaceName string, path string) {
+	// URL is supposed to have the following format
+	// ->   /api/terminal/{namespace}/{workspace-name}/{path} < optional
+	// -> 0 / 1 /    2   /    3      /        4       / 5
+	segments := strings.SplitN(requestPath, "/", 6)
+	if len(segments) < 5 {
+		return false, "", "", ""
+	} else {
+		namespace = segments[3]
+		workspaceName = segments[4]
+		if len(segments) == 6 {
+			path = segments[5]
+		}
+		return true, namespace, workspaceName, path
+	}
+}
+
+//getBaseTerminalHost evaluates ideUrl from the specified workspace and extract host from it
+func (p *Proxy) getBaseTerminalHost(ws *unstructured.Unstructured) (*url.URL, error) {
+	ideUrl, success, err := unstructured.NestedString(ws.UnstructuredContent(), "status", "ideUrl")
+	if !success {
+		return nil, errors.New("the specified workspace does not have ideUrl in its status")
+	}
+	if err != nil {
+		return nil, errors.New("failed to evaluate ide URL for the specified workspace. Cause: " + err.Error())
+	}
+
+	terminalUrl, err := url.Parse(ideUrl)
+	if err != nil {
+		return nil, errors.New("Failed to parse workspace ideUrl " + ideUrl)
+	}
+
+	terminalHost, err := url.Parse(terminalUrl.Scheme + "://" + terminalUrl.Host)
+	if err != nil {
+		return nil, errors.New("Failed to parse workspace ideUrl host " + ideUrl)
+	}
+
+	return terminalHost, nil
+}
+
+func (p *Proxy) proxyToWorkspace(workspaceIdeHost *url.URL, path string, token string, r *http.Request, w http.ResponseWriter) {
+	r2 := new(http.Request)
+	*r2 = *r
+	r2.URL = new(url.URL)
+	*r2.URL = *r.URL
+
+	r2.Header.Set("X-Forwarded-Access-Token", token)
+
+	r2.URL.Path = path
+
+	//TODO a new proxy per request is created. Must be revised and probably changed
+	terminalProxy := proxy.NewProxy(&proxy.Config{
+		Endpoint:        workspaceIdeHost,
+		TLSClientConfig: p.TLSClientConfig,
+	})
+
+	terminalProxy.ServeHTTP(w, r2)
+}
+
+//createDynamicClient create dynamic client with the configured token to be used
+func (p *Proxy) createDynamicClient(token string) (dynamic.Interface, error) {
+	tlsClientConfig := rest.TLSClientConfig{}
+	tlsClientConfig.Insecure = true
```

Comment:
If this is the endpoint is a service, use the injected service-ca. We definitely need to remove insecure: true before this can merge

Reply:
It should be fixed now. But currently, I've tested only off-cluster mode.
I'll resolve this comment when I test in-cluster mode as well.